geps/gep-1651.md

```diff
@@ -0,0 +1,268 @@
+this would lead to a combinatorial explosion as new dimensions (ie. IPv6) are
+introduced.
+
+From: https://github.com/kubernetes-sigs/gateway-api/pull/1653#issuecomment-1451246877
+
+> Although this makes sense in isolation, I'm worried about the long term impacts this could have. In my opinion, ClusterLocal is a modifier, not exactly an address type. For example, it's possible in the future that we'll have a way to provision cluster-local DNS names, we may want to use the same kind of mechanism to request a ClusterLocal DNS name for the Gateway.
+>
+> It's also possible that users will want to explicitly request an IP Families (v4, v6, or both). I'd really hate to get into a situation where we have the following options:
+>
+>   IPAddress
+>   IPv4Address
+>   IPv6Address
+>   ClusterLocalIPAddress
+>   ClusterLocalIPv4Address
+>   ClusterLocalIPv6Address
+>
+> For each dimension we avoid adding a separate field for and instead try to embed into a single name, we risk this kind of name explosion. Of course none of the above even begins to cover my idea of NetworkLocal which could further complicate this.
+
+### Scope/reachability field on GatewayAddress
+
+This would allow Gateways to have multiple scopes.
+
+From: https://github.com/kubernetes-sigs/gateway-api/pull/1653#issuecomment-1486271913
+> The obvious application for multiple scopes seems to be saving on boilerplate, which is a win, but are there are any other advantages to allowing one Gateway to have multiple scopes?
+>
+> Multiple scopes Pros:
+>
+> Allows a single Gateway to express multiple networks, saving on needing to attach HTTPRoutes to multiple Gateways for each network scope.
+>
+> Multiple scopes Cons:
+>
+> Complicates the Gateway's purpose. Instead of one Gateway being one set of Listeners, now a Gateway is two sets of listeners that have a totally different scope (and presumably, security context). Personally, I'm also concerned how this will interact with other features like merging and preprovisioning that GEP-1867: Per-Gateway Infrastructure #1868 will allow.
+
+### Adding `reachabiliy` attribute to GatewayClass
+
+See [Prior Art - Multiple Gateways Classes](#multiple-gateway-classes)
+
+## Survey of Prior Art
+
+These alternatives are a survey of existing approaches to support cluster
+local Gateways. Most are implementation specific and are not portable.
+
+### Special annotation/label
+
+Istio let's you specify an annotation `networking.istio.io/service-type` to
+change the underlying Kubernetes Service type to make it a ClusterIP type.
+
+### Re-use of AddressType Hostname
+
+Istio let's you re-use existing Gateway deployments by setting the address
+type to `Hostname` and the value to the Istio ingress Kubernetes Service. If an
+operator configures the Istio deployment to support cluster local traffic a
+Gateway implementation can select it using the `HostName` attribute.
+
+### Multiple Gateway Classes
```

Comment:
One of the parts that I don't see covered here is mutability. So far everything on Gateway resources can be changed. On the other hand, GatewayClass resources are to be interpreted as a template/point in time reference for Gateway creation. This is important because there are some concepts that are very difficult to change in a non-destructive way. At least for cloud providers, I think it would be difficult/impossible to support a change to Gateway reachability. I believe this would probably be the first time we introduced a field to the Gateway resource that implementations struggled to support changes to.

Reply:
Can most change address? seems about the same

Reply:
Is there any prior art/discussion about `address` mutability?

I don't really have an issue if `reachability` was immutable - would be good to get input from others. It's worth adding a section

Reply:
I added a section on mutability